jax/scipy/optimize/line_search.py

```diff
@@ -0,0 +1,453 @@
+import jax.numpy as jnp
+import jax
+from jax.lax import while_loop, cond
+from typing import NamedTuple
+
+
+class LineSearchResults(NamedTuple):
+  failed: bool  # True if the strong Wolfe criteria were satisfied
+  nit: int  # Number of iterations
+  nfev: int  # Number of functions evaluations
+  ngev: int  # Number of gradients evaluations
+  k: int  # Number of iterations
+  a_k: float  # Step size
+  f_k: jnp.ndarray  # Final function value
+  g_k: jnp.ndarray  # Final gradient value
+
+
+def line_search_backtracking(value_and_gradient, position, direction, f_0=None, g_0=None, max_iterations=50, c1=1e-4,
+                             c2=0.9):
+  """
+  Performs an inexact line-search. It is a modified backtracking line search. Instead of reducing step size by a
+  number < 1 if Wolfe conditions are not met, we check the sign of  u = del_t restricted_func(t).
+  If u > 0 then we do normal backtrack, otherwise we search forward. Normal backtracking can fail to satisfy strong
+  Wolfe conditions. This extra step costs one extra gradient evaluation.
+
+  Inspired by the backtracking algorithim on Wright and Nocedal, 'Numerical Optimization', 1999, pg. 37
+
+  Args:
+      value_and_gradient: function and gradient
+      position: position to search from
+      direction: descent direction to search along
+      f_0: optionally give starting function value at position
+      g_0: optionally give starting gradient at position
+      max_iterations: maximum number of searches
+      c1, c2: Wolfe criteria numbers from above reference
+
+  Returns: LineSearchResults
+
+  """
+
+  def restricted_func(t):
+    return value_and_gradient(position + t * direction)
+
+  grad_restricted = jax.grad(lambda t: restricted_func(t)[0])
```

Comment:
Indeed, I noted that `grad` is computed via `value_and_grad`. There's actually no point where it uses the first argument and not the second. There is one case where it uses the second and not the first, but since `grad` is `value_and_grad` this makes no difference.

Reply:
Note: this would make a difference in the implementation of the `zoom` method with `cond`. However, I still have yet to profile that implementation. There are 5 conditionals in the naive implementation, though I can lower that I think.